src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -2050,22 +2043,24 @@ JoinRestrictionListExistsInContext(JoinRestriction *joinRestrictionInput,
 		 * If we're dealing with different queries, we shouldn't treat their
 		 * restrictions as being the same.
 		 */
-		if (joinRestriction->plannerInfo != joinRestrictionInput->plannerInfo)
+		if (contextJoinRestriction->plannerInfo != joinRestriction->plannerInfo)
 		{
 			continue;
 		}
 
 		/*
-		 * We check whether the restrictions in joinRestriction is a super set
-		 * of the restrictions in joinRestrictionInput in the sense that all the
+		 * We check whether the restrictions in contextJoinRestriction is a super
+		 * set of the restrictions in joinRestriction in the sense that all the
 		 * restrictions in the latter already exists in the former.
 		 *
 		 * Also, note that list_difference() returns a list that contains all the
 		 * cells in joinRestrictInfoList that are not in inputJoinRestrictInfoList.
 		 * Finally, each element in these lists is a pointer to RestrictInfo
 		 * structure, where equal() function is implemented for the struct.
 		 */
-		if (list_difference(joinRestrictInfoList, inputJoinRestrictInfoList) == NIL)
+		List *contextJoinRestrictInfoList = contextJoinRestriction->joinRestrictInfoList;
+		List *joinRestrictInfoList = joinRestriction->joinRestrictInfoList;
+		if (LeftListIsSubset(contextJoinRestrictInfoList, joinRestrictInfoList))
```

Comment:
now reason behind https://github.com/citusdata/citus/issues/4094 is more obvious :)